test/parachain/tests/integration_test.rs

```diff
@@ -0,0 +1,404 @@
+
+// TODO: this is necessary for the jsonrpsee macro used
+#![allow(unused_variables, dead_code)]
+
+use assert_cmd::cargo::cargo_bin;
+use codec::Encode;
+use polkadot_primitives::parachain::{Info, Scheduling};
+use polkadot_primitives::Hash as PHash;
+use polkadot_runtime::{Header, OnlyStakingAndClaims, Runtime, SignedExtra, SignedPayload};
+use polkadot_runtime_common::{parachains, registrar, BlockHashCount};
+use serde_json::Value;
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_runtime::generic;
+use sp_version::RuntimeVersion;
+use std::io::Read;
+use std::{
+	convert::TryInto,
+	env, fs, io, net,
+	path::PathBuf,
+	process::{Child, Command, Stdio},
+	thread,
+	time::Duration,
+};
+use substrate_test_runtime_client::AccountKeyring::Alice;
+use tempfile::tempdir;
+
+static POLKADOT_ARGS: &[&str] = &["polkadot", "--chain=res/polkadot_chainspec.json"];
+
+jsonrpsee::rpc_api! {
+	Author {
+		#[rpc(method = "author_submitExtrinsic", positional_params)]
+		fn submit_extrinsic(extrinsic: String) -> PHash;
+	}
+
+	Chain {
+		#[rpc(method = "chain_getFinalizedHead")]
+		fn current_block_hash() -> PHash;
+
+		#[rpc(method = "chain_getHeader", positional_params)]
+		fn header(hash: PHash) -> Option<Header>;
+
+		#[rpc(method = "chain_getBlockHash", positional_params)]
+		fn block_hash(hash: Option<u64>) -> Option<PHash>;
+	}
+
+	State {
+		#[rpc(method = "state_getRuntimeVersion")]
+		fn runtime_version() -> RuntimeVersion;
+	}
+
+	System {
+		#[rpc(method = "system_networkState")]
+		fn network_state() -> Value;
+	}
+}
+
+// Adapted from
+// https://github.com/rust-lang/cargo/blob/485670b3983b52289a2f353d589c57fae2f60f82/tests/testsuite/support/mod.rs#L507
+fn target_dir() -> PathBuf {
+	env::current_exe()
+		.ok()
+		.map(|mut path| {
+			path.pop();
+			if path.ends_with("deps") {
+				path.pop();
+			}
+			path
+		})
+		.unwrap()
+}
+
+struct ChildHelper<'a> {
+	name: String,
+	child: &'a mut Child,
+	stdout: String,
+	stderr: String,
+}
+
+impl<'a> Drop for ChildHelper<'a> {
+	fn drop(&mut self) {
+		let name = self.name.clone();
+
+		self.terminate();
+		eprintln!(
+			"process '{}' stdout:\n{}\n",
+			name,
+			self.read_stdout_to_end().unwrap_or_default()
+		);
+		eprintln!(
+			"process '{}' stderr:\n{}\n",
+			name,
+			self.read_stderr_to_end().unwrap_or_default()
+		);
+	}
+}
+
+impl<'a> ChildHelper<'a> {
+	fn new(name: &str, child: &'a mut Child) -> ChildHelper<'a> {
+		ChildHelper {
+			name: name.to_string(),
+			child,
+			stdout: Default::default(),
+			stderr: Default::default(),
+		}
+	}
+
+	fn read_stdout_to_end(&mut self) -> io::Result<&str> {
+		let mut output = String::new();
+
+		self.child
+			.stdout
+			.as_mut()
+			.ok_or_else(|| io::Error::new(io::ErrorKind::Other, "stdout not captured"))?
+			.read_to_string(&mut output)?;
+		self.stdout.push_str(output.as_str());
+
+		Ok(&self.stdout)
+	}
+
+	fn read_stderr_to_end(&mut self) -> io::Result<&str> {
+		let mut output = String::new();
```

Comment:
Do you see the similarities of this function to the one above? :D

Reply:
What do you want exactly? A macro?

Reply:
I'd imagine a function like this:

```rust
fn read_to_end<'a, r: Read>(reader: &mut Option<R>, store: &'a mut String) -> io::Result<&'a str> {
    let mut output = String::new():

    reader
        .as_mut()
        .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "not captured"))?
        .read_to_string(&mut output)?;
    store.push_str(output.as_str())

    Ok(&output)
}

fn read_stdout_to_end(&mut self) -> io::Result<&str> {
    Self::read_to_end(&mut self.child.stdout, &mut self.stdout)
}
```

Reply:
Actually I don't think I need those functions anymore since I don't need to keep track of what has been read already. I will remove both.

Reply:
I don't think it's resolved as the code is still duplicated. But I think it's fine as it